File: snippets/screens/flutter/tutorial_screen/2/tutorial_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class TutorialScreen extends StatefulWidget {
  const TutorialScreen({super.key});

  @override
  State<TutorialScreen> createState() => _TutorialScreenState();
}

class _TutorialScreenState extends State<TutorialScreen>
    with TickerProviderStateMixin {
  int _currentStep = 0;
  late AnimationController _pulseController;
  late Animation<double> _pulseAnimation;
  bool _showOverlay = true;

  final List<TutorialStep> _tutorialSteps = [
    TutorialStep(
      title: 'Navigation Menu',
      description: 'Tap the menu icon to access different sections of the app.',
      targetKey: GlobalKey(),
      position: TutorialPosition.topLeft,
    ),
    TutorialStep(
      title: 'Start a Quiz',
      description: 'Tap here to begin your AI intelligence assessment.',
      targetKey: GlobalKey(),
      position: TutorialPosition.center,
    ),
    TutorialStep(
      title: 'View Your Progress',
      description: 'Check your scores and track improvement over time.',
      targetKey: GlobalKey(),
      position: TutorialPosition.bottomRight,
    ),
    TutorialStep(
      title: 'Settings & Profile',
      description: 'Customize your experience and manage your profile.',
      targetKey: GlobalKey(),
      position: TutorialPosition.topRight,
    ),
  ];

  @override
  void initState() {
    super.initState();

    SystemChrome.setSystemUIOverlayStyle(
      const SystemUiOverlayStyle(
        statusBarColor: Colors.transparent,
        statusBarIconBrightness: Brightness.light,
      ),
    );

    _pulseController = AnimationController(
      duration: const Duration(milliseconds: 1500),
      vsync: this,
    );

    _pulseAnimation = Tween<double>(
      begin: 1.0,
      end: 1.2,
    ).animate(CurvedAnimation(
      parent: _pulseController,
      curve: Curves.easeInOut,
    ));

    _pulseController.repeat(reverse: true);
  }

  @override
  void dispose() {
    _pulseController.dispose();
    super.dispose();
  }

  void _nextStep() {
    if (_currentStep < _tutorialSteps.length - 1) {
      setState(() {
        _currentStep++;
      });
    } else {
      _completeTutorial();
    }
  }

  void _previousStep() {
    if (_currentStep > 0) {
      setState(() {
        _currentStep--;
      });
    }
  }

  void _skipTutorial() {
    _completeTutorial();
  }

  void _completeTutorial() {
    setState(() {
      _showOverlay = false;
    });
    Future.delayed(const Duration(milliseconds: 300), () {
      Navigator.of(context).pushReplacementNamed('/permissions');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          // Mock app interface
          _buildMockInterface(),
          // Tutorial overlay
          if (_showOverlay) _buildTutorialOverlay(),
        ],
      ),
    );
  }

  Widget _buildMockInterface() {
    return Container(
      decoration: const BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: [
            Color(0xFF6366F1),
            Color(0xFF8B5CF6),
          ],
        ),
      ),
      child: SafeArea(
        child: Column(
          children: [
            // App bar
            Padding(
              padding: const EdgeInsets.all(16.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Container(
                    key: _tutorialSteps[0].targetKey,
                    child: const Icon(
                      Icons.menu,
                      color: Colors.white,
                      size: 28,
                    ),
                  ),
                  const Text(
                    'AIQ',
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: 24,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  Container(
                    key: _tutorialSteps[3].targetKey,
                    child: const Icon(
                      Icons.person,
                      color: Colors.white,
                      size: 28,
                    ),
                  ),
                ],
              ),
            ),
            // Main content
            Expanded(
              child: Container(
                margin: const EdgeInsets.all(16),
                decoration: BoxDecoration(
                  color: Colors.white,
                  borderRadius: BorderRadius.circular(24),
                ),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Container(
                      key: _tutorialSteps[1].targetKey,
                      width: 120,
                      height: 120,
                      decoration: BoxDecoration(
                        color: const Color(0xFF6366F1),
                        borderRadius: BorderRadius.circular(60),
                      ),
                      child: const Icon(
                        Icons.play_arrow,
                        color: Colors.white,
                        size: 60,
                      ),
                    ),
                    const SizedBox(height: 24),
                    const Text(
                      'Start Your AIQ Test',
                      style: TextStyle(
                        fontSize: 24,
                        fontWeight: FontWeight.bold,
                        color: Color(0xFF1F2937),
                      ),
                    ),
                    const SizedBox(height: 40),
                    Container(
                      key: _tutorialSteps[2].targetKey,
                      padding: const EdgeInsets.all(16),
                      margin: const EdgeInsets.symmetric(horizontal: 24),
                      decoration: BoxDecoration(
                        color: const Color(0xFFF3F4F6),
                        borderRadius: BorderRadius.circular(16),
                      ),
                      child: Row(
                        children: const [
                          Icon(
                            Icons.trending_up,
                            color: Color(0xFF6366F1),
                          ),
                          SizedBox(width: 12),
                          Text(
                            'View Progress',
                            style: TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.w500,
                            ),
                          ),
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildTutorialOverlay() {
    final currentStep = _tutorialSteps[_currentStep];

    return Container(
      color: Colors.black.withOpacity(0.8),
      child: Stack(
        children: [
          // Spotlight effect (simplified)
          Positioned.fill(
            child: AnimatedBuilder(
              animation: _pulseAnimation,
              builder: (context, child) {
                return CustomPaint(
                  painter: SpotlightPainter(
                    targetKey: currentStep.targetKey,
                    pulseScale: _pulseAnimation.value,
                  ),
                );
              },
            ),
          ),
          // Tutorial content
          Positioned(
            bottom: 100,
            left: 24,
            right: 24,
            child: Container(
              padding: const EdgeInsets.all(24),
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(16),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisSize: MainAxisSize.min,
                children: [
                  Row(
                    children: [
                      Container(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 8,
                          vertical: 4,
                        ),
                        decoration: BoxDecoration(
                          color: const Color(0xFF6366F1),
                          borderRadius: BorderRadius.circular(12),
                        ),
                        child: Text(
                          '${_currentStep + 1}/${_tutorialSteps.length}',
                          style: const TextStyle(
                            color: Colors.white,
                            fontSize: 12,
                            fontWeight: FontWeight.w600,
                          ),
                        ),
                      ),
                      const Spacer(),
                      TextButton(
                        onPressed: _skipTutorial,
                        child: const Text(
                          'Skip',
                          style: TextStyle(
                            color: Color(0xFF6B7280),
                            fontWeight: FontWeight.w500,
                          ),
                        ),
                      ),
                    ],
                  ),
                  const SizedBox(height: 16),
                  Text(
                    currentStep.title,
                    style: const TextStyle(
                      fontSize: 20,
                      fontWeight: FontWeight.bold,
                      color: Color(0xFF1F2937),
                    ),
                  ),
                  const SizedBox(height: 8),
                  Text(
                    currentStep.description,
                    style: const TextStyle(
                      fontSize: 16,
                      color: Color(0xFF6B7280),
                      height: 1.4,
                    ),
                  ),
                  const SizedBox(height: 24),
                  Row(
                    children: [
                      if (_currentStep > 0)
                        Expanded(
                          child: OutlinedButton(
                            onPressed: _previousStep,
                            style: OutlinedButton.styleFrom(
                              foregroundColor: const Color(0xFF6B7280),
                              side: const BorderSide(
                                color: Color(0xFFE5E7EB),
                              ),
                              shape: RoundedRectangleBorder(
                                borderRadius: BorderRadius.circular(12),
                              ),
                            ),
                            child: const Text('Previous'),
                          ),
                        ),
                      if (_currentStep > 0) const SizedBox(width: 12),
                      Expanded(
                        child: ElevatedButton(
                          onPressed: _nextStep,
                          style: ElevatedButton.styleFrom(
                            backgroundColor: const Color(0xFF6366F1),
                            foregroundColor: Colors.white,
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(12),
                            ),
                          ),
                          child: Text(
                            _currentStep == _tutorialSteps.length - 1
                                ? 'Finish'
                                : 'Next',
                          ),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class TutorialStep {
  final String title;
  final String description;
  final GlobalKey targetKey;
  final TutorialPosition position;

  TutorialStep({
    required this.title,
    required this.description,
    required this.targetKey,
    required this.position,
  });
}

enum TutorialPosition {
  topLeft,
  topRight,
  bottomLeft,
  bottomRight,
  center,
}

class SpotlightPainter extends CustomPainter {
  final GlobalKey targetKey;
  final double pulseScale;

  SpotlightPainter({
    required this.targetKey,
    required this.pulseScale,
  });

  @override
  void paint(Canvas canvas, Size size) {
    // Simplified spotlight effect
    final paint = Paint()
      ..color = Colors.white.withOpacity(0.2)
      ..style = PaintingStyle.fill;

    // Draw a simple circle for demonstration
    canvas.drawCircle(
      Offset(size.width / 2, size.height / 3),
      60 * pulseScale,
      paint,
    );
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}

File: snippets/screens/swiftui/tutorial_screen/1/tutorial_screen.swift
import SwiftUI

struct TutorialScreen: View {
    @State private var currentStep = 0
    @State private var showOverlay = true
    @State private var pulseScale: CGFloat = 1.0

    private let tutorialSteps: [TutorialStep] = [
        TutorialStep(
            title: "Navigation Menu",
            description: "Tap the menu icon to access different sections of the app.",
            position: .topLeft,
            targetFrame: CGRect(x: 16, y: 60, width: 28, height: 28)
        ),
        TutorialStep(
            title: "Start a Quiz",
            description: "Tap here to begin your AI intelligence assessment.",
            position: .center,
            targetFrame: CGRect(x: 150, y: 350, width: 120, height: 120)
        ),
        TutorialStep(
            title: "View Your Progress",
            description: "Check your scores and track improvement over time.",
            position: .bottomRight,
            targetFrame: CGRect(x: 24, y: 550, width: 200, height: 60)
        ),
        TutorialStep(
            title: "Settings & Profile",
            description: "Customize your experience and manage your profile.",
            position: .topRight,
            targetFrame: CGRect(x: 350, y: 60, width: 28, height: 28)
        )
    ]

    var body: some View {
        ZStack {
            // Mock app interface
            mockAppInterface

            // Tutorial overlay
            if showOverlay {
                tutorialOverlay
            }
        }
        .onAppear {
            startPulseAnimation()
        }
    }

    private var mockAppInterface: some View {
        VStack {
            // App bar
            HStack {
                Image(systemName: "line.horizontal.3")
                    .font(.system(size: 28))
                    .foregroundColor(.white)

                Spacer()

                Text("AIQ")
                    .font(.system(size: 24, weight: .bold))
                    .foregroundColor(.white)

                Spacer()

                Image(systemName: "person.circle")
                    .font(.system(size: 28))
                    .foregroundColor(.white)
            }
            .padding(.horizontal, 16)
            .padding(.top, 20)

            // Main content
            VStack {
                Spacer()

                VStack(spacing: 24) {
                    // Play button
                    ZStack {
                        Circle()
                            .fill(Color(red: 0.39, green: 0.40, blue: 0.95))
                            .frame(width: 120, height: 120)

                        Image(systemName: "play.fill")
                            .font(.system(size: 60))
                            .foregroundColor(.white)
                    }

                    Text("Start Your AIQ Test")
                        .font(.system(size: 24, weight: .bold))
                        .foregroundColor(Color(red: 0.12, green: 0.16, blue: 0.22))

                    // Progress card
                    HStack {
                        Image(systemName: "chart.line.uptrend.xyaxis")
                            .foregroundColor(Color(red: 0.39, green: 0.40, blue: 0.95))

                        Text("View Progress")
                            .font(.system(size: 16, weight: .medium))
                    }
                    .padding()
                    .frame(maxWidth: .infinity)
                    .background(Color(red: 0.95, green: 0.96, blue: 0.97))
                    .cornerRadius(16)
                    .padding(.horizontal, 24)
                }

                Spacer()
            }
            .background(Color.white)
            .clipShape(RoundedRectangle(cornerRadius: 24))
            .padding(.horizontal, 16)
            .padding(.bottom, 16)
        }
        .background(
            LinearGradient(
                gradient: Gradient(colors: [
                    Color(red: 0.39, green: 0.40, blue: 0.95),
                    Color(red: 0.55, green: 0.36, blue: 0.96)
                ]),
                startPoint: .top,
                endPoint: .bottom
            )
        )
    }

    private var tutorialOverlay: some View {
        ZStack {
            // Dark overlay
            Color.black.opacity(0.8)
                .ignoresSafeArea()

            // Spotlight effect
            SpotlightView(
                targetFrame: tutorialSteps[currentStep].targetFrame,
                pulseScale: pulseScale
            )

            // Tutorial content
            VStack {
                Spacer()

                TutorialContentCard(
                    step: tutorialSteps[currentStep],
                    currentStep: currentStep,
                    totalSteps: tutorialSteps.count,
                    onPrevious: previousStep,
                    onNext: nextStep,
                    onSkip: skipTutorial
                )
                .padding(.horizontal, 24)
                .padding(.bottom, 100)
            }
        }
    }

    private func startPulseAnimation() {
        withAnimation(.easeInOut(duration: 1.5).repeatForever(autoreverses: true)) {
            pulseScale = 1.2
        }
    }

    private func nextStep() {
        if currentStep < tutorialSteps.count - 1 {
            withAnimation(.easeInOut(duration: 0.3)) {
                currentStep += 1
            }
        } else {
            completeTutorial()
        }
    }

    private func previousStep() {
        if currentStep > 0 {
            withAnimation(.easeInOut(duration: 0.3)) {
                currentStep -= 1
            }
        }
    }

    private func skipTutorial() {
        completeTutorial()
    }

    private func completeTutorial() {
        withAnimation(.easeOut(duration: 0.3)) {
            showOverlay = false
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            // Navigate to permissions
            // NavigationManager.shared.navigate(to: .permissions)
        }
    }
}

struct TutorialContentCard: View {
    let step: TutorialStep
    let currentStep: Int
    let totalSteps: Int
    let onPrevious: () -> Void
    let onNext: () -> Void
    let onSkip: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 24) {
            // Header
            HStack {
                Text("\(currentStep + 1)/\(totalSteps)")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(Color(red: 0.39, green: 0.40, blue: 0.95))
                    .cornerRadius(12)

                Spacer()

                Button("Skip", action: onSkip)
                    .font(.system(size: 14, weight: .medium))
                    .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
            }

            // Content
            VStack(alignment: .leading, spacing: 8) {
                Text(step.title)
                    .font(.system(size: 20, weight: .bold))
                    .foregroundColor(Color(red: 0.12, green: 0.16, blue: 0.22))

                Text(step.description)
                    .font(.system(size: 16))
                    .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                    .lineSpacing(2)
            }

            // Buttons
            HStack(spacing: 12) {
                if currentStep > 0 {
                    Button(action: onPrevious) {
                        Text("Previous")
                            .font(.system(size: 14, weight: .medium))
                            .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                            .frame(maxWidth: .infinity)
                            .frame(height: 44)
                            .overlay(
                                RoundedRectangle(cornerRadius: 12)
                                    .stroke(Color(red: 0.90, green: 0.91, blue: 0.92), lineWidth: 1)
                            )
                    }
                }

                Button(action: onNext) {
                    Text(currentStep == totalSteps - 1 ? "Finish" : "Next")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .frame(height: 44)
                        .background(Color(red: 0.39, green: 0.40, blue: 0.95))
                        .cornerRadius(12)
                }
            }
        }
        .padding(24)
        .background(Color.white)
        .cornerRadius(16)
    }
}

struct SpotlightView: View {
    let targetFrame: CGRect
    let pulseScale: CGFloat

    var body: some View {
        // Simplified spotlight effect
        Circle()
            .fill(Color.white.opacity(0.2))
            .frame(width: 80 * pulseScale, height: 80 * pulseScale)
            .position(x: targetFrame.midX, y: targetFrame.midY)
    }
}

struct TutorialStep {
    let title: String
    let description: String
    let position: TutorialPosition
    let targetFrame: CGRect
}

enum TutorialPosition {
    case topLeft, topRight, bottomLeft, bottomRight, center
}

#Preview {
    TutorialScreen()
}